L82-core-event-engine.md

```diff
@@ -0,0 +1,478 @@
+  virtual TaskHandle Run(Callback fn, RunOptions opts) = 0;
+  /// Synonymous with scheduling an alarm to run at time \a when.
+  ///
+  /// The callback \a fn will execute when either when time \a when arrives
+  /// (receiving status OK), or when the \a fn is cancelled (reveiving status
+  /// CANCELLED). The callback is guaranteed to be called exactly once.
+  virtual TaskHandle RunAt(absl::Time when, Callback fn, RunOptions opts) = 0;
+  /// Immediately tries to cancel a callback.
+  /// Note that this is a "best effort" cancellation. No guarantee is made that
+  /// the callback will be cancelled, the call could be in any stage.
+  ///
+  /// There are three scenarios in which we may cancel a scheduled function:
+  ///   1. We cancel the execution before it has run.
+  ///   2. The callback has already run.
+  ///   3. We can't cancel it because it is "in flight".
+  ///
+  /// In all cases, the cancellation is still considered successful, the
+  /// callback will be run exactly once from either cancellation or from its
+  /// activation.
+  virtual void TryCancel(TaskHandle handle) = 0;
+```
+
+### Cleanup and Shutdown
+
+```cpp
+  /// Immediately run all callbacks with status indicating the shutdown. Every
+  /// EventEngine is expected to shut down exactly once. No new callbacks/tasks
+  /// should be scheduled after shutdown has begun, no new connections should be
+  /// created.
+  ///
+  /// If the \a on_shutdown_complete callback is given a non-OK status, errors
+  /// are expected to be unrecoverable. For example, an implementation could
+  /// warn callers about leaks if memory cannot be freed within a certain
+  /// timeframe.
+  virtual void Shutdown(Callback on_shutdown_complete) = 0;
+};
+
+}  // namespace experimental
+}  // namespace grpc_event_engine
+```
+
+### Application APIs for providing custom EventEngines
+
+*Caveat: This is not final, feedback is welcome!*
+
+There are three ways in which EventEngines can be provided to gRPC.
+
+#### Setting an EventEngine instance for a Channel
+
+EventEngines can be configured per channel, which allows customization for both
+clients and servers. To set a custom EventEngine for a client channel, you can
+do something like the following:
+
+```cpp
+ChannelArguments args;
+std::shared_ptr<EventEngine> engine = std::make_shared<MyEngine>(...);
+args.SetEventEngine(engine);
+MyAppClient client(grpc::CreateCustomChannel(
+    "localhost:50051", grpc::InsecureChannelCredentials(), args));
+```
+
+#### Setting an EventEngine instance for a Server
+
+A gRPC server can use a custom EventEngine by calling the
+`ServerBuilder::SetEventEngine` method:
+
+```cpp
+ServerBuilder builder;
+std::shared_ptr<EventEngine> engine = std::make_shared<MyEngine>(...);
+builder.SetEventEngine(engine);
+std::unique_ptr<Server> server(builder.BuildAndStart());
+server->Wait();
+```
+
+#### Providing a default EventEngine factory
+
+TODO: this has not been thoroughly discussed.
```

Comment:
I'd still prefer to see no default event engine in core - no concept of it at all in fact!

Instead, move this cost out to the wrapped languages and allow them to bind in an implementation that makes sense to them.

Doing so means that we can keep core *core* - nothing extraneous - and also likely helps push us to better point solutions in wrapped languages down the track.

Reply:
Maybe this section is talking about providing an API to let users set a default factory for this process?

Reply:
@lidizheng that's right.

@ctiller I see your argument. That's not really related to this section. I think you're engaging a much bigger discussion which might obviate this gRFC. Let's have that discussion between team leads and review an alternative proposal if you'd like to put one together. I can't really estimate how big a change it would be in the team's direction and investment over the past few quarters.

Reply:
Fair enough. I don't think it'd massively change direction to be fair, although it might add a step or two before we remove experimental, and I imagine first steps as experimental apis will necessarily have a default event engine - it's a sensible first step, and it'd be very hard to expunge iomgr without it.

My larger question is if it's a sensible end state, or if we end up with a more robust core of we omit the concept of a default event engine as part of that core.

I don't think we need to make this decision this week, but we should have a deeper discussion on this prior to grfc acceptance.

Reply:
It is a bit clunky. We do, however, need some solution for I/O to occur "behind the scenes", in a way that applications can control if they wish. Either that, or applications would pay the price of instantiating whatever mechanism gRPC provides for the background I/O - likely some default EventEngine in the current design, with its threads, memory usage, etc..

For example, last I checked, the Global XDS client channel has no surface API that would allow custom EventEngines to be provided. I think there are a handful such situations where some application-provided global default would be helpful.